File: test/features/ocr/time_extractor_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:war2aty/features/ocr/domain/services/time_extractor.dart';

void main() {
  const extractor = TimeExtractor();

  group('TimeExtractor', () {
    group('24h format', () {
      test('HH:MM', () {
        final results = extractor.extract('الساعة 14:30');
        expect(results, hasLength(1));
        expect(results.first.hour, equals(14));
        expect(results.first.minute, equals(30));
        expect(results.first.isAmbiguous, isFalse);
      });

      test('HH.MM', () {
        final results = extractor.extract('الموعد 09.00');
        expect(results, hasLength(1));
        expect(results.first.hour, equals(9));
        expect(results.first.minute, equals(0));
      });

      test('13:00-23:59 are unambiguous', () {
        final results = extractor.extract('13:00');
        expect(results.first.isAmbiguous, isFalse);
      });

      test('00:00 is unambiguous', () {
        final results = extractor.extract('00:00');
        expect(results.first.isAmbiguous, isFalse);
      });

      test('1:00-12:59 without marker are ambiguous', () {
        final results = extractor.extract('9:30');
        expect(results.first.isAmbiguous, isTrue);
      });
    });

    group('Arabic AM/PM', () {
      test('صباحاً (AM)', () {
        final results = extractor.extract('9:30 صباحاً');
        expect(results, hasLength(1));
        expect(results.first.hour, equals(9));
        expect(results.first.minute, equals(30));
        expect(results.first.isAmbiguous, isFalse);
      });

      test('مساءً (PM)', () {
        final results = extractor.extract('3:00 مساءً');
        expect(results, hasLength(1));
        expect(results.first.hour, equals(15));
        expect(results.first.minute, equals(0));
      });

      test('ص abbreviation (AM)', () {
        final results = extractor.extract('10:00 ص');
        expect(results, hasLength(1));
        expect(results.first.hour, equals(10));
      });

      test('م abbreviation (PM)', () {
        final results = extractor.extract('7:45 م');
        expect(results, hasLength(1));
        expect(results.first.hour, equals(19));
        expect(results.first.minute, equals(45));
      });

      test('12 PM stays 12', () {
        final results = extractor.extract('12:00 مساءً');
        expect(results.first.hour, equals(12));
      });

      test('12 AM becomes 0', () {
        final results = extractor.extract('12:00 صباحاً');
        expect(results.first.hour, equals(0));
      });
    });

    group('English AM/PM', () {
      test('AM', () {
        final results = extractor.extract('9:30 AM');
        expect(results, hasLength(1));
        expect(results.first.hour, equals(9));
      });

      test('PM', () {
        final results = extractor.extract('3:00 PM');
        expect(results.first.hour, equals(15));
      });

      test('a.m.', () {
        final results = extractor.extract('10:00 a.m.');
        expect(results.first.hour, equals(10));
      });

      test('p.m.', () {
        final results = extractor.extract('5:30 p.m.');
        expect(results.first.hour, equals(17));
      });
    });

    group('edge cases', () {
      test('rejects invalid minute > 59', () {
        final results = extractor.extract('10:60');
        expect(results, isEmpty);
      });

      test('rejects hour > 23 without AM/PM', () {
        final results = extractor.extract('25:00');
        expect(results, isEmpty);
      });

      test('multiple times in one text', () {
        final results = extractor.extract('من 9:00 صباحاً إلى 5:00 مساءً');
        expect(results, hasLength(2));
        expect(results[0].hour, equals(9));
        expect(results[1].hour, equals(17));
      });
    });
  });
}

File: supabase/tests/unit/time-extractor.test.ts
/**
 * F13-T05 · Tests for the time extractor.
 *
 * Mirrors `test/features/ocr/time_extractor_test.dart` case-for-case.
 */

import { assertEquals } from "jsr:@std/assert@1";

import { extractTimes } from "../../functions/_shared/extractors/time-extractor.ts";

// ── 24h format ───────────────────────────────────────────────────────────────

Deno.test("time: HH:MM", () => {
  const results = extractTimes("الساعة 14:30");
  assertEquals(results.length, 1);
  assertEquals(results[0].hour, 14);
  assertEquals(results[0].minute, 30);
  assertEquals(results[0].is_ambiguous, false);
});

Deno.test("time: HH.MM", () => {
  const results = extractTimes("الموعد 09.00");
  assertEquals(results.length, 1);
  assertEquals(results[0].hour, 9);
  assertEquals(results[0].minute, 0);
});

Deno.test("time: 13:00-23:59 are unambiguous", () => {
  assertEquals(extractTimes("13:00")[0].is_ambiguous, false);
});

Deno.test("time: 00:00 is unambiguous", () => {
  assertEquals(extractTimes("00:00")[0].is_ambiguous, false);
});

Deno.test("time: 1:00-12:59 without marker are ambiguous", () => {
  assertEquals(extractTimes("9:30")[0].is_ambiguous, true);
});

// ── Arabic AM/PM ─────────────────────────────────────────────────────────────

Deno.test("time: صباحاً (AM)", () => {
  const results = extractTimes("9:30 صباحاً");
  assertEquals(results.length, 1);
  assertEquals(results[0].hour, 9);
  assertEquals(results[0].minute, 30);
  assertEquals(results[0].is_ambiguous, false);
});

Deno.test("time: مساءً (PM)", () => {
  const results = extractTimes("3:00 مساءً");
  assertEquals(results.length, 1);
  assertEquals(results[0].hour, 15);
  assertEquals(results[0].minute, 0);
});

Deno.test("time: ص abbreviation (AM)", () => {
  const results = extractTimes("10:00 ص");
  assertEquals(results.length, 1);
  assertEquals(results[0].hour, 10);
});

Deno.test("time: م abbreviation (PM)", () => {
  const results = extractTimes("7:45 م");
  assertEquals(results.length, 1);
  assertEquals(results[0].hour, 19);
  assertEquals(results[0].minute, 45);
});

Deno.test("time: 12 PM stays 12", () => {
  assertEquals(extractTimes("12:00 مساءً")[0].hour, 12);
});

Deno.test("time: 12 AM becomes 0", () => {
  assertEquals(extractTimes("12:00 صباحاً")[0].hour, 0);
});

// ── English AM/PM ────────────────────────────────────────────────────────────

Deno.test("time: AM", () => {
  const results = extractTimes("9:30 AM");
  assertEquals(results.length, 1);
  assertEquals(results[0].hour, 9);
});

Deno.test("time: PM", () => {
  assertEquals(extractTimes("3:00 PM")[0].hour, 15);
});

Deno.test("time: a.m.", () => {
  assertEquals(extractTimes("10:00 a.m.")[0].hour, 10);
});

Deno.test("time: p.m.", () => {
  assertEquals(extractTimes("5:30 p.m.")[0].hour, 17);
});

// ── edge cases ───────────────────────────────────────────────────────────────

Deno.test("time: rejects invalid minute > 59", () => {
  assertEquals(extractTimes("10:60"), []);
});

Deno.test("time: rejects hour > 23 without AM/PM", () => {
  assertEquals(extractTimes("25:00"), []);
});

Deno.test("time: multiple times in one text", () => {
  const results = extractTimes("من 9:00 صباحاً إلى 5:00 مساءً");
  assertEquals(results.length, 2);
  assertEquals(results[0].hour, 9);
  assertEquals(results[1].hour, 17);
});
